lib/UserBackend.php

```diff
@@ -148,10 +149,63 @@ public function createUserIfNotExists($uid, array $attributes = []) {
 			}
 			$qb->execute();
 
+			// If we use per-user encryption the keys must be initialized first
+			$userSecret = $this->getUserSecret($uid, $attributes);
+			if ($userSecret !== null) {
+				$this->updateUserSecretHash($uid, $userSecret);
+				// Emit a post login action to initialize the encryption module with the user secret provided by the idp.
+				\OC_Hook::emit('OC_User', 'post_login', ['run' => true, 'uid' => $uid, 'password' => $userSecret, 'isTokenLogin' => false]);
+			}
 			$this->initializeHomeDir($uid);
 		}
 	}
 
+	private function getUserSecretHash($uid) {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('token')
+			->from('user_saml_auth_token')
+			->where($qb->expr()->eq('uid', $qb->createNamedParameter($uid)))
+			->andWhere($qb->expr()->eq('name', $qb->createNamedParameter('sso_secret_hash')))
+			->setMaxResults(10);
+		$result = $qb->execute();
+		$data = $result->fetchAll();
+		$result->closeCursor();
+		return $data;
+	}
+
+	private function checkUserSecretHash($uid, $userSecret) {
+		$data = $this->getUserSecretHash($uid);
+		foreach($data as $row) {
+			$storedHash = $row['token'];
+			if (\OC::$server->getHasher()->verify($userSecret, $storedHash, $newHash)) {
```

Comment:
please use ' \OCP\Server::get(OCP\Security\IHasher::class)` over deprecated and private `\OC::$server->getHasher()`. Also on in the other places.